tracer/src/Datadog.Trace/TraceProcessors/TraceUtil.cs

```diff
@@ -0,0 +1,344 @@
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Text;
+using Datadog.Trace.Util;
+
+namespace Datadog.Trace.TraceProcessors
+{
+    internal class TraceUtil
+    {
+        // MaxTagLength the maximum length a tag can have
+        private const int MaxTagLength = 200;
+
+        private static readonly UTF8Encoding Encoding = new UTF8Encoding(false);
+
+        // https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/traceutil/truncate.go#L36-L51
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool TruncateUTF8(ref string value, int limit)
+        {
+            if (string.IsNullOrEmpty(value) || Encoding.GetByteCount(value) <= limit)
+            {
+                return false;
+            }
+
+            var valueCharArray = value.ToCharArray();
+            for (var i = 0; i < valueCharArray.Length; i++)
+            {
+                if (Encoding.GetByteCount(valueCharArray, 0, i) > limit)
+                {
+                    value = value.Substring(0, i - 1);
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        // https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/agent/normalizer.go#L214-L219
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsValidStatusCode(string statusCode)
+        {
+            if (int.TryParse(statusCode, out int code))
+            {
+                return 100 <= code && code < 600;
+            }
+
+            return false;
+        }
+
+        // https://github.com/DataDog/datadog-agent/blob/eac2327c5574da7f225f9ef0f89eaeb05ed10382/pkg/trace/traceutil/normalize.go#L98-L209
+        public static string NormalizeTag(string value)
+        {
+            if (string.IsNullOrEmpty(value) || Encoding.GetByteCount(value) == 0)
+            {
+                return string.Empty;
+            }
+
+            char[] charArray = null;
+            char[] upArray = null;
+            char[] lowArray = null;
+            int trim = 0;
+            List<int[]> cuts = null;
+            int chars = 0;
+            int jump = 0;
+            int i = 0;
+
+            for (; i < value.Length; i++)
+            {
+                var c = value[i];
+                jump = 1;
+
+                if ((c >= 'a' && c <= 'z') || c == ':')
+                {
+                    chars++;
+                    goto end;
+                }
+
+                if (c >= 'A' && c <= 'Z')
+                {
+                    if (charArray is null)
+                    {
+                        charArray = value.ToCharArray();
+                    }
+
+                    charArray[i] = (char)(((int)value[i]) + ((int)'a' - (int)'A'));
+                    chars++;
+                    goto end;
+                }
+
+                if (char.IsUpper(c))
```

Comment:
We should benchmark of course, but I think we found that using code like this was faster than `char.IsUpper()`, `char.IsLetter()`, and `char.IsDigit()`. cc @andrewlock 

https://github.com/DataDog/dd-trace-dotnet/blob/af7cd30735c5ad0daf21291c4b238b7d1130fd29/tracer/src/Datadog.Trace/Util/UriHelpers.cs#L125-L129

Reply:
Yes, I'm alreading doing the fast path above:

https://github.com/DataDog/dd-trace-dotnet/blob/c669f0981cfbcd30ba37607a8686b90d11cc7f26/tracer/src/Datadog.Trace/TraceProcessors/TraceUtil.cs#L83

this is the fallback for those `unicode` chars that has an UpperCase representation outside the ASCII map.

Reply:
Sorry, I missed that. Thanks!